src/System.IO.Ports/src/System/IO/Ports/SerialPort.cs

```diff
@@ -0,0 +1,1307 @@
+        // control block and associated Win32 structures, and begins the event-watching cycle.
+        public SerialPort(string portName, int baudRate, Parity parity, int dataBits, StopBits stopBits)
+        {
+            PortName = portName;
+            BaudRate = baudRate;
+            Parity = parity;
+            DataBits = dataBits;
+            StopBits = stopBits;
+        }
+
+        // Calls internal Serial Stream's Close() method on the internal Serial Stream.
+        public void Close()
+        {
+            Dispose();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                if (IsOpen)
+                {
+                    _internalSerialStream.Flush();
+                    _internalSerialStream.Close();
+                    _internalSerialStream = null;
+                }
+            }
+            base.Dispose(disposing);
+        }
+
+        public void DiscardInBuffer()
+        {
+            if (!IsOpen)
+                throw new InvalidOperationException(SR.Port_not_open);
+            _internalSerialStream.DiscardInBuffer();
+            _readPos = _readLen = 0;
+        }
+
+        public void DiscardOutBuffer()
+        {
+            if (!IsOpen)
+                throw new InvalidOperationException(SR.Port_not_open);
+            _internalSerialStream.DiscardOutBuffer();
+        }
+
+        public static string[] GetPortNames()
+        {
+            // Hitting the registry for this isn't the only way to get the ports.
+            // 
+            // WMI: https://msdn.microsoft.com/en-us/library/aa394413.aspx
+            // QueryDosDevice: https://msdn.microsoft.com/en-us/library/windows/desktop/aa365461.aspx
+            //
+            // QueryDosDevice involves finding any ports that map to \Device\Serialx (call with null to get all, then iterate to get the actual device name)
+
+            RegistryKey baseKey = null;
+            RegistryKey serialKey = null;
+
+            String[] portNames = null;
+
+            try
+            {
+                baseKey = Registry.LocalMachine;
+                serialKey = baseKey.OpenSubKey(@"HARDWARE\DEVICEMAP\SERIALCOMM", false);
+
+                if (serialKey != null)
+                {
+
+                    string[] deviceNames = serialKey.GetValueNames();
+                    portNames = new String[deviceNames.Length];
+
+                    for (int i = 0; i < deviceNames.Length; i++)
+                        portNames[i] = (string)serialKey.GetValue(deviceNames[i]);
+                }
+            }
+            finally
+            {
+                baseKey?.Dispose();
+                serialKey?.Dispose();
+            }
+
+            // If serialKey didn't exist for some reason
+            if (portNames == null)
+                portNames = new String[0];
+
+            return portNames;
```

Comment:
Follow-up cleanup to consider at some point:

 - `using (RegistryKey serialKey = Registry.LocalMachine.OpenSubKey(@"HARDWARE\DEVICEMAP\SERIALCOMM"))`
 - Consider reusing the `string[]` returned from `serialKey.GetValueNames()` instead of allocating a new one
 -  `return portNames ?? Array.Empty<string>()`